analytic_engine/src/sst/parquet/writer.rs

```diff
@@ -149,16 +152,48 @@ impl RecordBatchGroupWriter {
     ) -> Result<RowGroupFilter> {
         let mut builder = RowGroupFilterBuilder::new(row_group_batch[0].schema_with_key());
 
+        let mut total_num_rows = 0;
         for partial_batch in row_group_batch {
-            for (col_idx, column) in partial_batch.columns().iter().enumerate() {
-                for row in 0..column.num_rows() {
-                    let datum = column.datum(row);
-                    let bytes = datum.to_bytes();
-                    builder.add_key(col_idx, &bytes);
+            total_num_rows += partial_batch.num_rows();
+        }
+        let mut digests = Vec::with_capacity(total_num_rows);
+
+        for (col_idx, col) in row_group_batch[0].columns().iter().enumerate() {
+            if !builder.has_builder(col_idx) {
+                builder.push_none();
+            } else {
+                digests.clear();
+                match col.datum_kind() {
+                    DatumKind::String=> {
+                        for partial_batch in row_group_batch {
+                            let column = partial_batch.column(col_idx);
+                            let arrow_array = column.to_arrow_array_ref();
+                            let string_array = arrow_array.as_string::<i32>();
+                            for row in 0..column.num_rows() {
+                                if string_array.is_null(row) {
+                                    digests.push(null_hash());
+                                } else {
+                                    digests.push(string_hash(string_array.value(row)));
+                                }
+                            }
+                        }
+                    }
+                    _ => {
+                        for partial_batch in row_group_batch {
+                            let column = partial_batch.column(col_idx);
+                            for row in 0..column.num_rows() {
+                                let datum = column.datum(row);
+                                let bytes = datum.to_bytes();
+                                digests.push(bytes_hash(&bytes));
+                            }
+                        }
+                    }
                 }
+                digests.sort_unstable();
```

Comment:
https://github.com/Fraser999/Hash-Hasher

Would you mind try this crate? It's designed for already hashed keys.